core/vgmgr/virtual_group_manager.go

```diff
@@ -56,6 +56,31 @@ func (p *PickVGFFilter) Check(vgfID uint32) bool {
 	return ok
 }
 
+type PickVGFByGVGFilter struct {
+	BlackListSPs map[uint32]struct{}
+}
+
+func NewPickVGFByGVGFilter(blackListSPs []uint32) *PickVGFByGVGFilter {
+	idSet := make(map[uint32]struct{})
+	for _, id := range blackListSPs {
+		idSet[id] = struct{}{}
+	}
+	return &PickVGFByGVGFilter{BlackListSPs: idSet}
+}
+
+// Check checks if a family has at least 1 valid GVG(gvg doest not contain blacklist Sp)
+func (p *PickVGFByGVGFilter) Check(gvgs map[uint32]*GlobalVirtualGroupMeta) bool {
+	validCount := len(gvgs)
+	for _, gvg := range gvgs {
+		for _, sspID := range gvg.SecondarySPIDs {
+			if _, ok := p.BlackListSPs[sspID]; ok {
+				validCount--
```

Comment:
if you find one black second sp, you should break the internal loop, other wise valid count could be a negative number.